### PR TITLE
Expose implementations of platform specific identities

### PR DESCRIPTION
**Pull request by shawn-hurley**
Currently putting this into the broker package, I thought this was an okay place because callers already need to import the broker package to get the request context struct. Implementation is based on the [spec definition](https://github.com/openservicebrokerapi/servicebroker/blob/master/profile.md#originating-identity-header).
(edit: including the link to OSB API Definitions)

### CODE REVIEW

File: pkg/broker/user_info.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"encoding/json"
+	"fmt"
+
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+)
+
+// ParseKubernetesIdentity - creates a kubernetes identity from the
+// orginating identity
+func ParseKubernetesIdentity(o osb.OriginatingIdentity) (*K8SIdentity, error) {
+	if o.Platform != "kubernetes" {
```
Comment (pmorie) [inline pkg/broker/user_info.go]: I made a PR to add constants for these to the client library: https://github.com/pmorie/go-open-service-broker-client/pull/111

File: pkg/broker/user_info.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	return &CloudFoundryIdentity{Platform: o.Platform, Value: m}, nil
+}
+
+// K8SIdentity - Identity implementation based on the kubernetes user info object
+type K8SIdentity struct {
+	Platform string
+	UserInfo UserInfo
+}
+
+// UserInfo - kubernetest user info object
```
Comment (pmorie) [inline pkg/broker/user_info.go]: typo

File: pkg/broker/user_info.go (hunk 1 above)
Comment (shawn-hurley) [reply pkg/broker/user_info.go]: I actually have a quick question about where the types are. To me, it seems that we should either have a shared place for the types or the types should be in the lib. It is mostly about where I would expect to find the types if I was looking. Just my two cents, and was wondering if that was a thought and not what we started with, or if you actually thought that was the wrong way to think about it?

Comment (pmorie) [reply pkg/broker/user_info.go]: I think the types should be here, because these are for consumption on the broker side. On the client side (in service-catalog), we serialize a k8s `AuthInfo` directly to get the value we send on the client request. Adding a k8s identity in go-open-service-broker-client doesn't really help us there.

We can always add the type here and move it to the client later if there's a demand for it.

File: pkg/broker/user_info.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+)
+
+// ParseKubernetesIdentity - creates a kubernetes identity from the
+// orginating identity
+func ParseKubernetesIdentity(o osb.OriginatingIdentity) (*K8SIdentity, error) {
+	if o.Platform != "kubernetes" {
+		return nil, fmt.Errorf("invalid plateform value, expecting kubernetes")
```
[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/broker/user_info.go]: Would it be helpful to maybe include in the error msg which platform it was?

File: pkg/broker/user_info.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		return nil, fmt.Errorf("unable to unmarshal json for value")
+	}
+	return &K8SIdentity{UserInfo: u, Platform: o.Platform}, nil
+}
+
+// ParseCloudFoundryIdentity - creates a cloud foundry identity from the
+// orginating identity
+func ParseCloudFoundryIdentity(o osb.OriginatingIdentity) (*CloudFoundryIdentity, error) {
+	if o.Platform != "cloudfoundry" {
+		return nil, fmt.Errorf("invalid platform value, expecting cloudfoundry")
```
Comment (LiliC) [inline pkg/broker/user_info.go]: Same as above: Would it be helpful to maybe include in the error msg which platform it was?

File: pkg/broker/user_info.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+// ParseKubernetesIdentity - creates a kubernetes identity from the
+// orginating identity
+func ParseKubernetesIdentity(o osb.OriginatingIdentity) (*K8SIdentity, error) {
+	if o.Platform != "kubernetes" {
+		return nil, fmt.Errorf("invalid plateform value, expecting kubernetes")
+	}
+
+	u := UserInfo{}
+	err := json.Unmarshal([]byte(o.Value), &u)
```
Comment (LiliC) [inline pkg/broker/user_info.go]: I noticed in the `ParseCloudFoundryIdentity` we validate if the `user_id` is present or not. Are there similar requirements for when the platform is K8s?

File: pkg/broker/user_info_test.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		{
+			name: "kubernetes invalid value json",
+			o: osb.OriginatingIdentity{
+				Platform: "kubernetes",
+				Value:    `{ser_id": "123"}`,
+			},
+			shouldErr: true,
+		},
+		{
+			name: "ok kubernetes identity",
```
Comment (LiliC) [inline pkg/broker/user_info_test.go]: Nit: `"kubernetes valid identity"` :)

File: pkg/broker/user_info_test.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+func TestParseCloudFoundryIdentity(t *testing.T) {
+	cases := []struct {
+		name              string
+		o                 osb.OriginatingIdentity
+		shouldErr         bool
+		expectingIdentity CloudFoundryIdentity
+	}{
+		{
+			name: "Cloud Foundry invalid platform",
```
Comment (LiliC) [inline pkg/broker/user_info_test.go]: Nit: I would either go with always upper case or always lower case. :)

File: pkg/broker/user_info.go (hunk 5 above)
Comment (shawn-hurley) [reply pkg/broker/user_info.go]: This is what I was attempting to work off of: https://github.com/openservicebrokerapi/servicebroker/blob/master/profile.md#kubernetes-originating-identity-header

I initially thought that the `json.Unmarshal` would make sure that this is the correct object. 

@pmorie @LiliC should we add `nil` checks for `Extras` and `Groups` and make `Username` and `UID` type defined pointers to a string to test if they are `nil` or not as well?

File: pkg/broker/user_info.go (hunk 1 above)
Comment (shawn-hurley) [reply pkg/broker/user_info.go]: @pmorie re: pmorie/go-open-service-broker-client/pull/111 should I update the dependency in this PR or should another PR be opened to update vendor?

Comment (shawn-hurley) [reply pkg/broker/user_info.go]: Added PR: #9 to bump the client to get the new constants.

**Comment by shawn-hurley** [comment]
Depends on #9 to use the new constants in `v2/constants.go` from pmorie/go-open-service-broker-client#111

**Comment by LiliC** [comment]
@shawn-hurley Seem like that PR was merged, now you can bump up the dep on the client to `0.0.5` and then we can re-review this?

**Comment by shawn-hurley** [comment]
@LiliC If you look at #9 that has the code to bump the vendor. If that could be merged first then this should be good.

**Comment by LiliC** [comment]
@shawn-hurley Sorry about that, misread the comment above!

**Comment by shawn-hurley** [comment]
@LiliC @pmorie just a ping to let you know the dependent PR has been merged

File: pkg/broker/user_info.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"encoding/json"
+	"fmt"
+
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+)
+
+// ParseKubernetesIdentity - creates a kubernetes identity from the
+// orginating identity
+func ParseKubernetesIdentity(o osb.OriginatingIdentity) (*K8SIdentity, error) {
```
Comment (LiliC) [inline pkg/broker/user_info.go]: I would prefer to have this as a generic `ParseIdentity` function instead, so as a user I don't have to match platforms but that the lib resolves that for me and returns the correct Identity.

But can't currently come up with any sane suggestion, any ideas?

**Comment by LiliC** [comment]
Also another thing is would be great to have an example use case in the osb-starter-pack broker for this. WDYT?

File: pkg/broker/user_info.go (hunk 8 above)
Comment (shawn-hurley) [reply pkg/broker/user_info.go]: I thought of that as well, the issue that I keep coming up with is that you would need an interface to encapsulate the behavior between the two Identities. And they are different as when I get the k8s identity I want the UserInfo structure, and when I get cloud foundry I expect a map[string]interface{}. 

We could eventually do something like `GetValue(valueName string) interface{}` or `GetValueString(valueName) (string, error)` but I think that will be very complex and maybe less usable then dealing with calling different methods? 

@pmorie I think the above point is what you thought as well when we discussed my first attempt at this PR.

**Comment by shawn-hurley** [comment]
> to have an example use case in the osb-starter-pack broker for this

I think that makes sense. The question I have is should that be added to the starter-pack after this is added to the lib and is released or should we create the example now with a dependency on this PR? I think developing that might be harder than the first way IMO

**Comment by LiliC** [comment]
> I think that makes sense. The question I have is should that be added to the starter-pack after this is added to the lib and is released or should we create the example now with a dependency on this PR? I think developing that might be harder than the first way IMO

IMO after this is released should be fine.

File: pkg/broker/user_info.go (hunk 8 above)
Comment (pmorie) [reply pkg/broker/user_info.go]: We could have a union type `Identity` with fields `Kubernetes` and `CloudFoundry` and then an `Unknown` bucket for everything else.

Comment (shawn-hurley) [reply pkg/broker/user_info.go]: So `Kubernetes` = `UserInfo`, `CloudFoundry` = `map[string]interface{}`? So is unkown for a platform that we do not understand?

Comment (n3wscott) [reply pkg/broker/user_info.go]: Perhaps you could parse the identity and return a wrapped result that can return the type, and has known named methods to return typed k8s and CF identity objects, and maybe a method to also return the data back as a map? So the caller has to do some work to figure out which representation to ask for but it will get a typed object back without the weird union of the three things. 

so maybe like:

```
func ParseIdentity(o osb.OriginatingIdentity) (*Identity, error)

identity, err := ParseIdentity(osbIdent)

switch identity.Type {
case K8s: 
    k8sIdentity := identity.AsK8s()
case CF:
    cfIdentity := identity.AsCF()  
case Unknown:
    mapIdentity := identity.AsMap()
}

...
```

Comment (shawn-hurley) [reply pkg/broker/user_info.go]: I am fine with this approach if it sounds better? @LiliC @pmorie

Comment (LiliC) [reply pkg/broker/user_info.go]: The approach @n3wscott suggested sgtm 👍

Comment (LiliC) [reply pkg/broker/user_info.go]: > a quick question about that implementation what should the AsXX function do if it is called but the platform type does not match? should it return the zero value s for the typed struct? (or map)?

Can you rephrase that, not sure I understood? You mean if it's neither k8s nor CF?

> Should the validation of the type (CF check for key "user_id") be in the AsCF function or do you think that the ParseIdentity does that? Sorry just want to make sure everyone is on the same page for the next implementation of this one.

IMO I would return early so in the `ParseIdentity()` function already we should verify each platform before returning it, so in that case it would just return the error.

Comment (n3wscott) [reply pkg/broker/user_info.go]: > IMO I would return early so in the ParseIdentity() function already we should verify each platform before returning it, so in that case it would just return the error.

I agree with this. Internally it knows what the identity type is (because it also supports .Type, so it can throw an error if Type is k8s but you are asking for CF.)

Comment (n3wscott) [reply pkg/broker/user_info.go]: Btw, I am short handing the named for the As**, feel free to use AsKubernetes() etc.

Comment (shawn-hurley) [reply pkg/broker/user_info.go]: I will implement this way if you want, just know the way it is now is a simpler way for a caller to interact IMO then what we are proposing EX:

```golang
switch originatingIdentity.Platform {
case osb.KubernetesPlatform:
     k8sIdentity := ParseKubernetesIdentity(originatingIdentity)
     ....
case osb.CloudFoundryPlatform:
    cfIdentity := ParseCloudFoundryIdentity(originatingIdentity)
default:
    //decode to map[string]interface{} if you want.
     .....
}
```

VS
```
func ParseIdentity(o osb.OriginatingIdentity) (*Identity, error) {...}

identity, err := ParseIdentity(osbIdent)

switch identity.Type {
case K8s: 
    k8sIdentity := identity.AsK8s()
case CF:
    cfIdentity := identity.AsCF()  
case Unknown:
    mapIdentity := identity.AsMap()
}
```

I personally don't see the value in adding `identity.Type` and its constants. Adding an Identity struct attaching three new methods and having a client still have to do a switch statement just after making a call when the above would work. Just my $0.02

Comment (jmrodri) [reply pkg/broker/user_info.go]: If there were no switch in the second case I could see the value in it. But with both needing a case statement, I'm not sure I see why one is easier than the other. You still need to know am I running in k8s or cf.

Comment (kibbles-n-bytes) [reply pkg/broker/user_info.go]: I think the former is cleaner, but only because the assumption here is that the processing was actually already done to get us the `osb.OriginatingIdentity` resource. If you're able to switch off `originatingIdentity.Platform`, there's no need to further process it into a generic `Identity` struct to then switch off of.

Comment (pmorie) [reply pkg/broker/user_info.go]: Use a union type; a struct that has pointers to a k8s identity, cf identity, and other.

**Comment by shawn-hurley** [comment]
@LiliC @pmorie @n3wscott @kibbles-n-bytes @jmrodri 

I have updated the implementation if you all could take a look and see if this is what you were expecting.

File: pkg/broker/user_info.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	return identity, nil
+}
+
+// Identity - union type, used to access the correct identity implementation
+// type
+type Identity struct {
+	Platform     string
+	Kubernetes   *UserInfo
+	CloudFoundry map[string]interface{}
+	Unkown       map[string]interface{}
```
Comment (jmrodri) [inline pkg/broker/user_info.go]: typo: Unkown should be Unknown

File: pkg/broker/user_info.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		err := json.Unmarshal([]byte(o.Value), &m)
+		if err != nil {
+			return identity, fmt.Errorf("unable to unmarshal json for value")
+		}
+		identity.Unknown = m
+	}
+	return identity, nil
+}
+
+// Identity - union type, used to access the correct identity implementation
```
Comment (LiliC) [inline pkg/broker/user_info.go]: I would add the "Originating" word somewhere in there just for clarity as thats how the osb spec names it as well. Either in the struct name or at least in the doc.

**Comment by jmrodri** [comment]
LGTM now

File: pkg/broker/user_info.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+)
+
+// parseKubernetesIdentity - creates a kubernetes identity from the
+// orginating identity
+func parseKubernetesIdentity(o osb.OriginatingIdentity) (*UserInfo, error) {
+	u := UserInfo{}
+	err := json.Unmarshal([]byte(o.Value), &u)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal json for value")
```
Comment (LiliC) [inline pkg/broker/user_info.go]: Would suggest to include Kubernetes in this error, so something like this:

> unable to unmarshal json for value while parsing Kubernetes Identity.

WDYT?

File: pkg/broker/user_info.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	return &u, nil
+}
+
+// parseCloudFoundryIdentity - creates a cloud foundry identity from the
+// orginating identity
+func parseCloudFoundryIdentity(o osb.OriginatingIdentity) (map[string]interface{}, error) {
+	m := map[string]interface{}{}
+	err := json.Unmarshal([]byte(o.Value), &m)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal json for value")
```
Comment (LiliC) [inline pkg/broker/user_info.go]: Same as above, would be useful to include that it failed while parsing CF Identity in the error msg. WDYT?

File: pkg/broker/user_info.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	// Validate that user_id MUST be in the json object.
+	if _, ok := m["user_id"]; !ok {
+		return nil, fmt.Errorf("user_id key was not found in json object")
```
Comment (LiliC) [inline pkg/broker/user_info.go]: Same here... something like this would be useful, IMO:

> user_id key was not found in json cloud foundry object

File: pkg/broker/user_info.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	return identity, nil
+}
+
+// Identity - union type, used to access the correct identity implementation
+// type
+type Identity struct {
+	Platform     string
+	Kubernetes   *UserInfo
+	CloudFoundry map[string]interface{}
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply pkg/broker/user_info.go]: https://github.com/openservicebrokerapi/servicebroker/blob/master/profile.md#originating-identity-header

File: pkg/broker/user_info.go [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	return identity, nil
+}
+
+// Identity - union type, used to access the correct originating identity
+// implementation type
+type Identity struct {
+	Platform     string
+	Kubernetes   *UserInfo
+	CloudFoundry map[string]interface{}
```
Comment (LiliC) [inline pkg/broker/user_info.go]: I would say we create a struct similar to the `UserInfo` with the field `user_id` and `extra` or `leftover` field, that will hold all the leftovers when unmarshmaling the from the json. WDYT? This way we adhere to the spec but also have clear separation between data types.

File: pkg/broker/user_info.go [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+	Unknown      map[string]interface{}
+}
+
+// UserInfo - kubernetes user info object
+type UserInfo struct {
```
Comment (LiliC) [inline pkg/broker/user_info.go]: Considering this is just used for K8s, we should rename it to `KubernetesUserInfo`, wdyt?

**Comment by jmrodri** [comment]
lgtm, what's left here?
